toolchain/driver/driver.cpp

```diff
@@ -22,286 +25,481 @@
+  } else if (result == Args::ParseResult::MetaSuccess) {
+    return true;
   }
 
-  llvm::StringRef target_triple;
-  if (dump_mode == DumpMode::Assembly && !args.empty() &&
-      args.front().starts_with("--target_triple=")) {
-    target_triple = args.front().split("=").second;
-    args = args.drop_front();
-  }
-
-  llvm::StringRef output_file;
-  if (dump_mode == DumpMode::ObjectCode) {
-    while (!args.empty() && (args.front().starts_with("--target_triple=") ||
-                             args.front().starts_with("--output_file="))) {
-      if (args.front().starts_with("--target_triple=")) {
-        target_triple = args.front().split("=").second;
-        args = args.drop_front();
-      }
-      if (args.front().starts_with("--output_file=")) {
-        output_file = args.front().split("=").second;
-        args = args.drop_front();
-      }
-    }
-
-    if (output_file.empty()) {
-      error_stream_ << "ERROR: Must provide an output file.\n";
-      return false;
-    }
+  if (options.verbose) {
+    // Note this implies streamed output in order to interleave.
+    vlog_stream_ = &error_stream_;
   }
 
-  if (args.empty()) {
-    error_stream_ << "ERROR: No input file specified.\n";
-    return false;
+  switch (options.subcommand) {
+    case Options::Subcommand::Compile:
+      return Compile(options.compile_options);
   }
+  llvm_unreachable("All subcommands handled!");
+}
 
-  llvm::StringRef input_file_name = args.front();
-  args = args.drop_front();
-  if (!args.empty()) {
-    ReportExtraArgs("dump", args);
-    return false;
+auto Driver::Compile(const CompileOptions& options) -> bool {
+  StreamDiagnosticConsumer stream_consumer(error_stream_);
+  DiagnosticConsumer* consumer = &stream_consumer;
+  std::unique_ptr<SortingDiagnosticConsumer> sorting_consumer;
+  if (vlog_stream_ == nullptr && !options.stream_errors) {
+    sorting_consumer = std::make_unique<SortingDiagnosticConsumer>(*consumer);
+    consumer = sorting_consumer.get();
   }
 
-  CARBON_VLOG() << "*** SourceBuffer::CreateFromFile ***\n";
-  auto source = SourceBuffer::CreateFromFile(fs_, input_file_name);
+  CARBON_VLOG() << "*** SourceBuffer::CreateFromFile on '"
+                << options.input_file_name << "' ***\n";
+  auto source = SourceBuffer::CreateFromFile(fs_, options.input_file_name);
   CARBON_VLOG() << "*** SourceBuffer::CreateFromFile done ***\n";
   if (!source.ok()) {
     error_stream_ << "ERROR: Unable to open input source file: "
                   << source.error();
     return false;
   }
+  CARBON_VLOG() << "*** file:\n```\n" << source->text() << "\n```\n";
 
   bool has_errors = false;
 
+  using Phase = CompileOptions::Phase;
+  switch (options.phase) {
+    case Phase::Tokenize:
+      if (options.dump_parse_tree) {
+        error_stream_ << "ERROR: Requested dumping the parse tree but compile "
+                         "phase is limited to '"
+                      << options.phase << "'\n";
+        has_errors = true;
```

Comment:
Should these exit more immediately, because they're user-caused flag errors versus compile-time errors?

That is, I'm fine if you want to collect all of them, but I'm not convince the system should even open the file when there are invalid flags.

Also, maybe this would be better done as part of post-flag parsing validation? In the Options `Build`, could that handle this and fail really early?